cpp/dolfinx/mesh/Topology.cpp

```diff
@@ -702,28 +712,107 @@ std::vector<std::int32_t> convert_to_local_indexing(
 
 //-----------------------------------------------------------------------------
 Topology::Topology(MPI_Comm comm, CellType cell_type)
-    : _comm(comm), _cell_type(cell_type),
+    : _comm(comm), _index_map(cell_dim(cell_type) + 1, {nullptr}),
       _connectivity(
           cell_dim(cell_type) + 1,
           std::vector<std::shared_ptr<graph::AdjacencyList<std::int32_t>>>(
               cell_dim(cell_type) + 1))
 {
-  // Do nothing
+  std::int8_t tdim = cell_dim(cell_type);
+  _entity_type_offsets.resize(tdim + 2);
+  for (std::int8_t i = 0; i < tdim + 2; ++i)
+    _entity_type_offsets[i] = i;
+
+  _entity_types = {CellType::point, CellType::interval};
+  if (tdim == 2)
+    _entity_types.push_back(cell_type);
+  else if (tdim == 3)
+  {
+    _entity_types.push_back(cell_facet_type(cell_type, 0));
+    _entity_types.push_back(cell_type);
+  }
+}
+//-----------------------------------------------------------------------------
+Topology::Topology(MPI_Comm comm, const std::vector<CellType>& cell_types)
+    : _comm(comm), _entity_types({mesh::CellType::point}),
+      _entity_type_offsets({0, 1})
+{
+  assert(cell_types.size() > 0);
+  std::int8_t tdim = cell_dim(cell_types[0]);
+
+  if (tdim > 1)
+  {
+    // In 2D, the facet is an interval
+    _entity_types.push_back(CellType::interval);
+    _entity_type_offsets.push_back(_entity_types.size());
+
+    if (tdim == 3)
+    {
+      // Find all facet types
+      std::vector<mesh::CellType> facet_types;
+      for (auto c : cell_types)
+      {
+        assert(cell_dim(c) == tdim);
+        std::int8_t nf = cell_num_entities(c, tdim - 1);
+        for (std::int8_t i = 0; i < nf; ++i)
+          facet_types.push_back(cell_facet_type(c, i));
+      }
+      std::sort(facet_types.begin(), facet_types.end());
+      facet_types.erase(std::unique(facet_types.begin(), facet_types.end()),
+                        facet_types.end());
+      _entity_types.insert(_entity_types.end(), facet_types.begin(),
+                           facet_types.end());
+
+      _entity_type_offsets.push_back(_entity_types.size());
+    }
+  }
+
+  // Cell Types
+  _entity_types.insert(_entity_types.end(), cell_types.begin(),
+                       cell_types.end());
+  _entity_type_offsets.push_back(_entity_types.size());
+
+  std::int8_t conn_size = _entity_type_offsets.back();
+  _index_map.resize(conn_size);
+
+  // Create square list of lists
+  _connectivity.resize(conn_size);
+  for (auto& c : _connectivity)
+    c.resize(conn_size);
 }
 //-----------------------------------------------------------------------------
-int Topology::dim() const noexcept { return _connectivity.size() - 1; }
+int Topology::dim() const noexcept { return _entity_type_offsets.size() - 2; }
 //-----------------------------------------------------------------------------
 void Topology::set_index_map(int dim,
```

Comment:
Do we need to change the documentation of this function? as it now only sets the index map for the first entity of the given dimension?

Reply:
Probably should throw an error if there are multiple maps.

Reply:
Topology code should be almost exclusively `std::int32_t` or `std::int64_t` - avoid `int` unless it's obvious that it's wide enough.

Reply:
This one can be `std::int8_t`